File: dsa/7.linked_list/21_Merge_Two_Sorted_Lists.py
import typing as t
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: t.Optional[ListNode], list2: t.Optional[ListNode]) -> t.Optional[ListNode]:
        pointer1: ListNode = list1
        pointer2: ListNode = list2

        pointer: ListNode = ListNode()
        result: ListNode = pointer

        while pointer1 and pointer2:
            if pointer1.val < pointer2.val:
                pointer.next = pointer1
                pointer1 = pointer1.next
            else:
                pointer.next = pointer2
                pointer2 = pointer2.next

            pointer = pointer.next
        
        while pointer1:
            pointer.next = pointer1
            pointer1 = pointer1.next
            pointer = pointer.next
        
        while pointer2:
            pointer.next = pointer2
            pointer2 = pointer2.next
            pointer = pointer.next
        
        return result.next

File: dsa/7+8.linked_list/21_Merge_Two_Sorted_Lists.java
package linked_list;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode ind1 = list1;
        ListNode ind2 = list2;

        ListNode list = new ListNode(0);
        ListNode ind = list;

        while (ind1 != null & ind2 != null){
            if (ind1.val > ind2.val){
                ind.next = new ListNode(ind2.val);
                ind2 = ind2.next;
            }
            else {
                ind.next = new ListNode(ind1.val);
                ind1 = ind1.next;
            }
            ind = ind.next;
        }

        while (ind1 != null){
            ind.next = new ListNode(ind1.val);
            ind1 = ind1.next;
            ind = ind.next;
        }

        while (ind2 != null){
            ind.next = new ListNode(ind2.val);
            ind2 = ind2.next;
            ind = ind.next;
        }
        return list.next;
    }
}
